File: bango-test/src/main/java/sonique/bango/driver/panel/serviceproblem/ServiceProblemPanel.java
package sonique.bango.driver.panel.serviceproblem;

import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import sky.sns.spm.domain.model.refdata.PresentedServiceType;
import sky.sns.spm.domain.model.refdata.ServiceType;
import sky.sns.spm.domain.model.serviceproblem.ServiceProblemStatus;
import sonique.bango.driver.component.HasTitle;
import sonique.bango.driver.component.form.SupermanPanel;
import spm.domain.OperatorReference;
import spm.domain.QueueName;
import spm.domain.ServiceProblemId;
import spm.domain.SnsServiceId;
import spm.messages.bt.types.DirectoryNumber;

import java.time.LocalDateTime;
import java.util.Date;

import static sonique.utils.StringUtils.unCamel;

public class ServiceProblemPanel extends SupermanPanel implements HasTitle {
    public ServiceProblemPanel(ServiceProblemTabContent serviceProblemTabContent) {
        super(serviceProblemTabContent, By.cssSelector("[id^='serviceProblemPanel']"));
    }

    @Override
    public String title() {
        WebElement titleElement = element().findElement(By.cssSelector("div.x-title-text"));
        return titleElement.getText();
    }

    public ServiceProblemId serviceProblemId() {
        return new ServiceProblemId(textFieldValue("Service Problem Id"));
    }

    public SnsServiceId serviceId() {
        return new SnsServiceId(textFieldValue("Service Id"));
    }

    public ServiceProblemStatus status() {
        return ServiceProblemStatus.valueOf(textFieldValue("Status"));
    }

    public DirectoryNumber directoryNumber() {
        return new DirectoryNumber(textFieldValue("Directory No"));
    }

    public QueueName queue() {
        return new QueueName(textFieldValue("Queue"));
    }

    public LocalDateTime openedDate() {
        return dateField("Opened Date", "dd/MM/yyyy HH:mm").value();
    }

    public LocalDateTime closedDate() {
        return dateField("Closed Date", "dd/MM/yyyy HH:mm").value();
    }

    public PresentedServiceType serviceType() {
        return fromDescription(textFieldValue("Service Type"));
    }

    public String customerName() {
        return textFieldValue("Customer Name");
    }

    public String contactNumber() {
        return textFieldValue("Contact No");
    }

    public String operatorAccountNumber() {
        return textFieldValue("Chordiant Acc No");
    }

    public String problem() {
        return textFieldValue("Problem");
    }

    public OperatorReference operatorReference() {
        return new OperatorReference(textFieldValue("Operator Ref"));
    }

    public String fault() {
        return textFieldValue("Fault");
    }

    public String cause() {
        return textFieldValue("Cause");
    }

    public String resolutionReason() {
        return textFieldValue("Resolution Reason");
    }

    private String textFieldValue(String label) {
        return textField(label).value();
    }

    private PresentedServiceType fromDescription(String description) {
        for (PresentedServiceType serviceTypeCode : PresentedServiceType.values()) {
            if (serviceTypeCode.getDisplayName().equals(description)){
                return serviceTypeCode;
            }
        }
        throw new IllegalArgumentException(String.format("No %s for [%s]", unCamel(ServiceType.class), description));
    }
}

File: bango-js/src/main/javascript/app/view/serviceproblem/ServiceProblemPanel.js
Ext.define('Spm.view.serviceproblem.ServiceProblemPanel', {
    extend: 'Ext.panel.Panel',
    alias: 'widget.serviceProblemPanel',

    requires: [
        'Ext.container.Container',
        'Ext.form.FieldContainer',
        'Ext.form.field.Date',
        'Ext.form.field.Text',
        'Ext.layout.container.Form',
        'Ext.layout.container.HBox',
        'Ext.layout.container.VBox'
    ],

    layout: {
        align: 'stretch',
        type: 'vbox'
    },

    title: 'Service Problem',
    frame: true,
    fieldDefaults: {
        disabled: true
    },

    bind: {
        iconCls: '{assignedStateIconClass}'
    },

    items: [
        {
            xtype: 'container',
            layout: {
                align: 'stretch',
                type: 'hbox'
            },
            defaults: {
                layout: 'form',
                minWidth: 260,
                defaults: {readOnly:true}
            },
            items: [
                {
                    xtype: 'fieldcontainer',
                    flex: 1,
                    fieldDefaults: {
                        labelWidth: 110
                    },
                    items: [
                        {
                            xtype: 'textfield',
                            bind: {
                                value: '{serviceProblem.serviceProblemId}'
                            },
                            fieldLabel: 'Service Problem Id'
                        },
                        {
                            xtype: 'textfield',
                            bind: {
                                value: '{serviceProblem.snsServiceId}'
                            },
                            fieldLabel: 'Service Id'
                        },
                        {
                            xtype: 'textfield',
                            fieldLabel: 'Chordiant Acc No',
                            bind: {
                                value: '{serviceProblem.chordiantAccountNumber}'
                            }
                        }
                    ]
                },
                {
                    xtype: 'fieldcontainer',
                    flex: 1,
                    margin: '0 0 0 3',
                    items: [
                        {
                            xtype: 'textfield',
                            name: 'status',
                            bind: {
                                value: '{serviceProblem.status}'
                            },
                            fieldLabel: 'Status'
                        },
                        {
                            xtype: 'textfield',
                            fieldLabel: 'Service Type',
                            bind: {
                                value: '{serviceProblem.serviceType.description} ({serviceProblem.assignmentCode})'
                            }
                        },
                        {
                            xtype: 'textfield',
                            fieldLabel: 'Operator Ref',
                            bind: {
                                value: '{serviceProblem.operatorReference}'
                            }
                        }
                    ]
                },
                {
                    xtype: 'fieldcontainer',
                    flex: 1,
                    margin: '0 0 0 3',
                    items: [
                        {
                            xtype: 'datefield',
                            fieldLabel: 'Opened Date',
                            bind: {
                                value: '{serviceProblem.openedDate}'
                            },
                            format: 'd/m/y H:i',
                            hideTrigger: true
                        },
                        {
                            xtype: 'datefield',
                            fieldLabel: 'Closed Date',
                            bind: {
                                value: '{serviceProblem.closedDate}'
                            },
                            format: 'd/m/y H:i',
                            hideTrigger: true
                        }
                    ]
                },
                {
                    xtype: 'fieldcontainer',
                    flex: 1,
                    margin: '0 0 0 3',
                    items: [
                        {
                            xtype: 'textfield',
                            bind: {
                                value: '{serviceProblem.directoryNumber}'
                            },
                            fieldLabel: 'Directory No'
                        },
                        {
                            xtype: 'textfield',
                            fieldLabel: 'Customer Name',
                            bind: {
                                value: '{serviceProblem.customerName}'
                            }
                        },
                        {
                            xtype: 'textfield',
                            fieldLabel: 'Contact No',
                            bind: {
                                value: '{serviceProblem.contactNumber}'
                            }
                        }
                    ]
                }
            ]
        },
        {
            xtype: 'fieldcontainer',
            minWidth: 1050,
            fieldDefaults: {
                labelWidth: 110,
                readOnly:true
            },
            layout: 'form',
            items: [
                {
                    xtype: 'textfield',
                    fieldLabel: 'Queue',
                    bind: {
                        value: '{serviceProblem.queue.name}'
                    }
                },
                {
                    xtype: 'textfield',
                    fieldLabel: 'Problem',
                    bind: {
                        value: '{serviceProblem.problem}'
                    }
                },
                {
                    xtype: 'textfield',
                    fieldLabel: 'Fault',
                    bind: {
                        value: '{serviceProblem.fault}'
                    }
                },
                {
                    xtype: 'textfield',
                    fieldLabel: 'Cause',
                    bind: {
                        value: '{serviceProblem.cause}'
                    }
                },
                {
                    xtype: 'textfield',
                    fieldLabel: 'Resolution Reason',
                    bind: {
                        value: '{serviceProblem.resolutionReason}'
                    }
                }
            ]
        }
    ]

});
